eo-phi-normalizer/app/Main.hs

```diff
@@ -1,111 +1,222 @@
+      (Nothing, Just expression') -> pure expression'
+      _ -> die [i|You must specify either -#{head inputFileLongName}|--#{inputFileLongName} #{_FILE} or #{_PROGRAM}|]
+  case parseProgram src of
+    Left err -> die [i|"An error occurred parsing the input program: #{err}|]
+    Right program -> pure program
+
+getLoggers :: Maybe FilePath -> IO (String -> IO (), String -> IO ())
+getLoggers outputFile = do
+  handle <- maybe (pure stdout) (`openFile` WriteMode) outputFile
+  pure
+    ( \x -> hPutStrLn handle x >> hFlush handle
+    , \x -> hPutStr handle x >> hFlush handle
+    )
 
 main :: IO ()
 main = do
-  opts <- getRecord "Normalizer"
-  let (CLIOptions params inPath) = opts
-  let (CLINamedParams{..}) = params
-  case rulesYaml of
-    Just path -> do
-      handle <- maybe (pure stdout) (`openFile` WriteMode) outPath
-      let logStr = hPutStr handle
-      let logStrLn = hPutStrLn handle
-      ruleSet <- parseRuleSetFromFile path
+  opts <- customExecParser pprefs cliOpts
+  case opts of
+    CLI'MetricsPhi' CLI'MetricsPhi{..} -> do
+      let ?parserContext = Optparse.Context metricsCommandName metricsParserInfo
+      program' <- getProgram inputFile program
+      (logStrLn, _) <- getLoggers outputFile
+      let metrics = collectMetrics program'
+      logStrLn $ encodeToJSONString metrics
+    CLI'TransformPhi' CLI'TransformPhi{..} -> do
+      let ?parserContext = Optparse.Context transformCommandName transformParserInfo
+      program' <- getProgram inputFile program
+      (logStrLn, logStr) <- getLoggers outputFile
+      ruleSet <- parseRuleSetFromFile rulesPath
       unless (single || json) $ logStrLn ruleSet.title
-      src <- maybe getContents readFile inPath
-      let progOrError = parseProgram src
-      case progOrError of
-        Left err -> error ("An error occurred parsing the input program: " <> err)
-        Right input@(Program bindings) -> do
-          let results
-                | chain = applyRulesChain (Context (convertRule <$> ruleSet.rules) [Formation bindings]) (Formation bindings)
-                | otherwise = pure <$> applyRules (Context (convertRule <$> ruleSet.rules) [Formation bindings]) (Formation bindings)
-              uniqueResults = nub results
-              totalResults = length uniqueResults
-          when (null uniqueResults || null (head uniqueResults)) $ error "Could not normalize the program"
-          let printFormationAsProgramOrObject = \case
-                Formation bindings' -> printTree $ Program bindings'
-                x -> printTree x
-          if single
+      let Program bindings = program'
+          results
+            | chain = applyRulesChain (Common.Context (convertRule <$> ruleSet.rules) [Formation bindings]) (Formation bindings)
+            | otherwise = pure <$> applyRules (Common.Context (convertRule <$> ruleSet.rules) [Formation bindings]) (Formation bindings)
+          uniqueResults = nub results
+          totalResults = length uniqueResults
```

Comment:
Unrelated to this PR, but just so we don't forget: Let's avoid operations that don't work with infinite lists because then the program hangs indefinitely rather than printing anything at all.
We could also define `applyRulesN` to avoid infinite application of rules

Reply:
Oh, I shouldn't have committed that. It's not ready yet (needs a CLI option) and I didn't push `applyRulesN` itself 😅